### PR TITLE
RuntimeEvent kafka sink

### LINKED ISSUE #242: Implement Kafka sink for runtime events
Kafka messages should be sent on when SIMONA finishes (with or without error).

For this it would be convenient to be able to implement various sink types, similarly to how `ResultEventListener` handles events.
The `SimonaListener` concept can be adapted or discarded in this process.

### PR DESCRIPTION
**Pull request by sebastian-peter**
Resolves #242 

Also addresses a small part of #48

### CODE REVIEW

**Review by t-ober — CHANGES_REQUESTED** [review]
Looks very nice - thanks ! Just minor stuff.

File: src/main/scala/edu/ie3/simona/config/ConfigFailFast.scala [hunk 1]
```diff
@@ -263,6 +269,18 @@ case object ConfigFailFast extends LazyLogging {
...
+  /** Check the runtime event listener config
+    * @param listenerConfig
+    *   the runtime listener config
+    */
+  private def checkRuntimeListenerConfiguration(
+      listenerConfig: SimonaConfig.Simona.Runtime.Listener
+  ): Unit = {
+    listenerConfig.kafka.foreach(kafka =>
+      checkKafkaParams(kafka, Seq(kafka.topic))
+    )
```
[the remark this answers is not in the document]
Comment (sebastian-peter) [reply src/main/scala/edu/ie3/simona/config/ConfigFailFast.scala:281]: `eventsToProcess` can be `None`, in which case no filter is applied to received runtime events

File: src/main/scala/edu/ie3/simona/event/listener/RuntimeEventListener.scala [hunk 2]
```diff
@@ -6,130 +6,71 @@
...
-    val s = math pow (10, precision)
-    (math round number * s) / s
-  }
+  def apply(
+      listeners: Iterable[RuntimeEventSink],
+      eventsToProcess: Option[List[String]] = None
+  ): Behavior[RuntimeEvent] = Behaviors
+    .receive[RuntimeEvent] { case (ctx, event) =>
+      eventsToProcess match {
+        case None => processEvent(listeners, event, ctx.log)
```
[the remark this answers is not in the document]
Comment (sebastian-peter) [reply src/main/scala/edu/ie3/simona/event/listener/RuntimeEventListener.scala:53]: That's how it works?

File: src/main/scala/edu/ie3/simona/event/listener/RuntimeEventListener.scala [hunk 3]
```diff
@@ -6,130 +6,71 @@
...
-  }
-
-  private def calcTime(currentTick: Long): String = {
-    TimeUtil.withDefaults.toString(
-      currentTick.toDateTime(
-        TimeUtil.withDefaults.toZonedDateTime(startDateTimeString)
-      )
+      ),
+      listenerConf.kafka.map(kafkaConf => RuntimeEventKafkaSink(kafkaConf)),
+      queue.map(qu => RuntimeEventQueueSink(qu))
```
[the remark this answers is not in the document]
Comment (sebastian-peter) [reply src/main/scala/edu/ie3/simona/event/listener/RuntimeEventListener.scala:38]: It's currently only used in tests. I added some ScalaDoc to all sinks now

File: src/main/scala/edu/ie3/simona/io/runtime/RuntimeEventLogSink.scala [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        )
+
+      case Error(errMsg) =>
+        log.error(s"$errMsg")
+
+      case _ =>
+        log.error(
+          s"Unexpected run time event received: $runtimeEvent."
+        )
+    }
```
[the remark this answers is not in the document]
Comment (sebastian-peter) [reply src/main/scala/edu/ie3/simona/io/runtime/RuntimeEventLogSink.scala:64]: Check out `RuntimeEventListener#apply`. A log sink is always initialized and the queue sink is only initialized if a queue was supplied. Same thing with the kafka sink: Only if corresponding configuration is present, such a sink is initialized and used.
I added some documentation which could be useful here.

**Comment by sebastian-peter** [comment]
I think we should wait until we completed some integration tests with simona-service here. I don't currently don't foresee any problems here, but let's wait until that's tested.

**Comment by sebastian-peter** [comment]
The manual integration tests completed with success.

**Review by t-ober — DISMISSED** [review]

**Review by t-ober — APPROVED** [review]

**Pull request merged**